### PR TITLE
Assignment2

### PR DESCRIPTION
**Pull request by SScura**

### CODE REVIEW

File: SecretSanta/src/SecretSanta.Api/SecretSanta.Api.csproj [hunk 1]
```diff
@@ -8,6 +8,12 @@
   <ItemGroup>
     <Compile Include="..\..\GlobalSuppression.cs" Link="GlobalSuppression.cs" />
   </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.4.0" />
+    <PackageReference Include="MSTest.TestAdapter" Version="2.0.0" />
+    <PackageReference Include="MSTest.TestFramework" Version="2.0.0" />
+    <PackageReference Include="coverlet.collector" Version="1.2.0">
```
Comment (mmwoodfo) [inline SecretSanta/src/SecretSanta.Api/SecretSanta.Api.csproj]:
```suggestion
    <PackageReference Include="coverlet.collector" Version="1.2.0"/>
```

1>Build FAILED.
1>
1>...\SecretSanta\src\SecretSanta.Api\SecretSanta.Api.csproj(16,5): error MSB4025: The project file could not be loaded. The 'PackageReference' start tag on line 15 position 6 does not match the end tag of 'ItemGroup'. Line 16, position 5.
1>    0 Warning(s)
1>    1 Error(s)
1>
1>Time Elapsed 00:00:00.13

File: SecretSanta/src/SecretSanta.Web/SecretSanta.Web.csproj [hunk 2]
```diff
@@ -3,6 +3,11 @@
   <PropertyGroup>
     <TargetFramework>netcoreapp3.1</TargetFramework>
   </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="3.1.1" />
+    <PackageReference Include="Microsoft.Extensions.Logging" Version="3.1.1" />
+    <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="3.1.1" />
+  </ItemGroup>>
```
Comment (mmwoodfo) [inline SecretSanta/src/SecretSanta.Web/SecretSanta.Web.csproj]:
```suggestion
  </ItemGroup>
```
1>Build FAILED.
1>
1>...\SecretSanta\src\SecretSanta.Web\SecretSanta.Web.csproj(1,1): error MSB4067: The element <#text> beneath element <Project> is unrecognized.
1>    0 Warning(s)
1>    1 Error(s)
1>
1>Time Elapsed 00:00:00.05

File: SecretSanta/src/SecretSanta.Api/SecretSanta.Api.csproj (hunk 1 above)
Comment (SScura) [reply SecretSanta/src/SecretSanta.Api/SecretSanta.Api.csproj]: Darn! Good eyes and thank you for the tip on the diag settings!

Comment (mmwoodfo) [reply SecretSanta/src/SecretSanta.Api/SecretSanta.Api.csproj]: Yeah np! I can't figure out how to get the .dll message to go away, I've tried cleaning the solution and deleting the bin file to no avail. Not going to mess with it too much more, but I would be curious to hear how you get it to work, as I see it's still failing. ):

Comment (SScura) [reply SecretSanta/src/SecretSanta.Api/SecretSanta.Api.csproj]: So... I used your suggestion and changed the setting. Still failed on me but for some reason I didn't see any warnings, again. Then after I realized it failed I went back in and VS threw an error at me in the Program.cs file. I realized I had not removed the Intellitect Analyzers references. I cleaned up some proj file refs, did a clean/rebuild, and something in there resolved my issue. I'm pretty sure I missed removing those analyzers somewhere along the way.

Comment (mmwoodfo) [reply SecretSanta/src/SecretSanta.Api/SecretSanta.Api.csproj]: Ah okay! Glad you fixed it. Good to know, thanks for the update! :)

**Comment by breaman** [comment]
##  (SScura)

## Assignment2

| Project Analysis Summary |   |
| -----------------| - |
| Nullable Enabled: | **✔** |
| Unexpected Disabled Warnings:  | **✔**
| Appropriately handle all warnings: | **✔** |

| ProjectName | Nullable | Solutions | Analyzers | Unexpected Disabled Warnings | Waning Level |
| -- | -- | -- | -- | -- | -- |
| SecretSanta.Api | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Business | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Data | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Web | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Api.Tests | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Business.Tests | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Data.Tests | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Web.Tests | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |


## Main Assignment

1. Add the following classes to the **SecretSanta.Data** project with the following associated properties:
   - **`EntityBase`** ✔
     - `int Id`
   - **`FingerPrintEntityBase`** inherit from **`EntityBase`** ✔
     - `string CreatedBy`
     - `DateTime CreatedOn`
     - `string ModifiedBy`
     - `DateTime ModifiedOn`
   - **`Group`** inherit from **`FingerPrintEntityBase`** ✔
     - `string Name`
   - Create a class that will hold the many to many relationship between User and Group ✔
     - Add a collection on both User and Group that references this class
2. Modify the following classes in the **SecretSanta.Data** project:
   - **`Gift`** ✔
     - Remove the Id property and inherit from FingerPrintEntityBase
   - **`User`** ✔
     - Remove the Id property and inherit from FingerPrintEntityBase
     - Make `Gifts` read/write
     - Create a nullable Santa property that contains a User as someone's Santa
3. Remove non-default constructors for `Gift` and `User`. ✔
4. In the **SecretSanta.Data** project
   - Create the **`ApplicationDbContext`** class ✔
     - Create two constructors ✔
        - One only takes the DbContextOptions and passes it on to the base class
        - One that takes the DbContextOptions and an IHttpContextAccessor, pass the DbContextOptions to the base class and save off the IHttpContextAccessor in a class property
     - Add DbSet properties for User, Gift, and Group classes ✔
     - Override OnModelCreating method and use the fluent API to configure the many-to-many pieces that are needed for setting up the database properly ✔
     - Override both SaveChanges methods and apply the fingerprinting logic ✔
5. Update unit tests in the **`SecretSanta.Data.Tests`** project to test actual database logic instead of just constructor logic
   - Mock up IHttpContextAccessor to verify that fingerprint information is working correctly (using Moq library) ✔


## Extra Credit

The following are options for extra credit:
- Enable console logging that will log all database commands ✔

**Pull request closed without merging**